# origin: marcelo (BitKeeper)
# cset: 1.1560 (2.4) key=41e2ccd0OuVN0bKOhZvnda0zXqnTsA
# inclusion: upstream
# descrition: Brad Spengler: Fix RLIMIT_MEMLOCK issue
# revision date: Thu, 13 Jan 2005 15:12:37 +0900
#
# S rset: ChangeSet|1.1559..1.1560
# I rset: include/linux/mm.h|1.49..1.50
#
# Key:
# S: Skipped  ChangeSet file only
# O: Original Followed by Updated
# U: Updated  Included with updated range of versions
# I: Included Included verbatim
# E: Excluded Excluded on request from user
# D: Deleted  Manually deleted by subsequent user edit
#
#
# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2005/01/10 16:43:28-02:00 marcelo@logos.cnet 
#   Brad Spengler: Fix RLIMIT_MEMLOCK issue
# 
# include/linux/mm.h
#   2005/01/10 16:41:47-02:00 marcelo@logos.cnet +8 -0
#   Brad Spengler: Fix RLIMIT_MEMLOCK issue
#   ,
# 
#
===== include/linux/mm.h 1.49 vs 1.50 =====
--- 1.49/include/linux/mm.h	2005-01-07 20:14:01 +09:00
+++ 1.50/include/linux/mm.h	2005-01-11 03:41:47 +09:00
@@ -660,6 +660,14 @@
 		spin_unlock(&vma->vm_mm->page_table_lock);
 		return -ENOMEM;
 	}
+
+	if ((vma->vm_flags & VM_LOCKED) &&
+      	    ((vma->vm_mm->locked_vm + grow) << PAGE_SHIFT) > current->rlim[RLIMIT_MEMLOCK].rlim_cur) {
+		spin_unlock(&vma->vm_mm->page_table_lock);
+		return -ENOMEM;
+	}
+
+
 	vma->vm_start = address;
 	vma->vm_pgoff -= grow;
 	vma->vm_mm->total_vm += grow;
